File: algorithm/java/src/main/java/com/liyongquan/math/MissingTwo.java
package com.liyongquan.math;

/**
 * 面试题 17.19. 消失的两个数字
 * <p>
 * 给定一个数组，包含从 1 到 N 所有的整数，但其中缺了两个数字。你能在 O(N) 时间内只用 O(1) 的空间找到它们吗？
 * <p>
 * 以任意顺序返回这两个数字均可。
 * <p>
 * 示例 1:
 * <p>
 * 输入: [1]
 * 输出: [2,3]
 * 示例 2:
 * <p>
 * 输入: [2,3]
 * 输出: [1,4]
 * 提示：
 * <p>
 * nums.length <= 30000
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/missing-two-lcci
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class MissingTwo {
    public int[] missingTwo(int[] nums) {
        int len = nums.length;
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        int sum2 = (len + 2) * (len + 3) / 2;
        int diff = sum2 - sum;
        //缺失的两个数必然在divide的两侧，剩下的问题就是找缺失的一个数字
        int divide = diff / 2;
        int sum3 = 0;
        for (int num : nums) {
            if (num <= divide) {
                sum3 += num;
            }
        }
        int a = (1 + divide) * divide / 2 - sum3;
        return new int[]{a, diff - a};
    }
}

File: algorithm/golang/math/missingTwo.go
package math

import "sort"

// 给定一个数组，包含从 1 到 N 所有的整数，但其中缺了两个数字。你能在 O(N) 时间内只用 O(1) 的空间找到它们吗？
//
//以任意顺序返回这两个数字均可。
//
//示例 1:
//
//输入: [1]
//输出: [2,3]
//示例 2:
//
//输入: [2,3]
//输出: [1,4]
//提示：
//
//nums.length <= 30000
//
//来源：力扣（LeetCode）
//链接：https://leetcode.cn/problems/missing-two-lcci
//著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

func missingTwo(nums []int) []int {
	sort.Ints(nums)
	n := len(nums) + 2
	sum := 0
	for _, num := range nums {
		sum += num
	}
	diff := (n+1)*n/2 - sum
	c := diff / 2
	// 找到第一个>c的下标
	var search func(c int) int
	search = func(c int) int {
		l := 0
		r := len(nums) - 1
		for l < r {
			mid := l + (r-l)/2
			if nums[mid] > c {
				r = mid
			} else {
				l = mid + 1
			}
		}
		return l
	}
	split := search(c)
	//[0,split),[split,n-2)
	s1 := 0
	for i := 0; i < split; i++ {
		s1 += nums[i]
	}
	nums1 := (split+1)*(split)/2 - s1
	nums2 := diff - nums1
	return []int{nums1, nums2}
}
